JDK-8312395: Improve assertions in growableArray

MBaesken opened the pull request:
There are a number of assertions in growableArray , for example to check for a valid index to access/remove elements.
Those assertions can be improved, e.g. by showing the bad index value used in case of failure.

Example for an assertion in the 'at(index i)' access method 
old assertion looks like
assert(0 <= i && i < _len) failed: illegal index

new assertion looks like
assert(0 <= i && i < _len) failed: illegal index -559030609, 1 accessible elements

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8312395](https://bugs.openjdk.org/browse/JDK-8312395): Improve assertions in growableArray (**Bug** - P4)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [d4e3ba68](https://git.openjdk.org/jdk/pull/14946/files/d4e3ba687413ec1dd93535eab55bd042b288d634)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [d4e3ba68](https://git.openjdk.org/jdk/pull/14946/files/d4e3ba687413ec1dd93535eab55bd042b288d634)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14946/head:pull/14946` \
`$ git checkout pull/14946`

Update a local copy of the PR: \
`$ git checkout pull/14946` \
`$ git pull https://git.openjdk.org/jdk.git pull/14946/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14946`

View PR using the GUI difftool: \
`$ git pr show -t 14946`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14946.diff">https://git.openjdk.org/jdk/pull/14946.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14946#issuecomment-1643446540)

dholmes-ora commented on src/hotspot/share/utilities/growableArray.hpp:
```diff
@@ -142,17 +142,17 @@ class GrowableArrayView : public GrowableArrayBase {
 
   E& at(int i) {
-    assert(0 <= i && i < _len, "illegal index");
+    assert(0 <= i && i < _len, "illegal index %d, %d accessible elements", i, _len);
```
How about:

"Illegal index %d for length %d"

?

MBaesken replied:
That sounds like a good suggestions !

dholmes-ora reviewed (changes requested):
Approval in principle but I have a suggestion.

Thanks

MBaesken commented:
I adjusted the output following your suggestion and also added some output to a few more asserts.

Btw any idea why so much 'this->_len' syntax is used at some places and not just _len or other field names ?

tstuefe reviewed (approved):
Looks good!

tstuefe commented:
looks 

> I adjusted the output following your suggestion and also added some output to a few more asserts.
> 
> Btw any idea why so much 'this->_len' syntax is used at some places and not just _len or other field names ?

Probably some IDE-code-completion-related artifacts. E.g. typing "this->" in CDT gives me all members, and maybe someone just hit Enter then and took the full completion

dholmes-ora commented on src/hotspot/share/utilities/growableArray.hpp:
```diff
@@ -259,8 +259,8 @@ class GrowableArrayView : public GrowableArrayBase {
-    assert(0 <= start, "illegal index");
-    assert(start < end && end <= _len, "erase called with invalid range");
+    assert(0 <= start, "illegal start index %d", start);
+    assert(start < end && end <= _len, "erase called with invalid range %d, %d", start, end);
```
I think you'd want to see `_len` here too - suggestion:

"erase called with invalid range (%d, %d) for length %d", start, end, _len

MBaesken replied:
Hi David, I agree !

dholmes-ora reviewed (approved):
One further suggestion but approving anyway.

I think the use of `this->xxx` is just personal style. I find it unnecessary especially given the other uses in this file.

Thanks.

tstuefe commented:
> 
> I think the use of `this->xxx` is just personal style. I find it unnecessary especially given the other uses in this file.
> 

Interestingly enough I get compiler errors when leaving out "this->" (GCC 10.3). It may be because the member is part of a templatized base class. Odd, though.

> Thanks.

MBaesken commented:
Hi David and Thomas, thanks for the reviews !

/integrate

kimbarrett commented:
> > I think the use of `this->xxx` is just personal style. I find it unnecessary especially given the other uses in this file.
> 
> Interestingly enough I get compiler errors when leaving out "this->" (GCC 10.3). It may be because the member is part of a templatized base class. Odd, though.

Qualification with `this->` is needed in a bunch of places here because of the
C++ name lookup rules. Unqualified name lookup does not search dependent base
classes. For example, in GrowableArrayWithAllocator an unqualified use of
`_len` won't find that member in the base class, but will instead (probably)
fail to find it in namespace scope during phase1 (pre-instantiation) template
checking lookup (of 2 phase lookup), and so fail to compile. It isn't possible
to resolve it in phase1 because the dependent base class definition isn't
known until instantiation.

dholmes-ora commented:
Thanks @kimbarrett !
